### ISSUE TITLE
Add Index on JSON Field

### ISSUE THREAD

**Issue by AzSiAz** [mikro-orm/mikro-orm#1230]
**Is your feature request related to a problem? Please describe.**
You can't add an index on json field like you would do in sql
```sql
CREATE INDEX "test_on_json_index" ON test((data->>'somefield'), "userId");
```

**Describe the solution you'd like**
A small update to `@Index()` to be able to do something like this
```ts
import { BaseEntity, Entity, Index, JsonType, PrimaryKey, Property } from '@mikro-orm/core'

@Index({ properties: ["data->>'somefield'", 'userId'] })
@Entity()
export class Test extends BaseEntity<Test, 'id'> {
    @PrimaryKey({ columnType: 'uuid', defaultRaw: 'uuid_generate_v4()' })
    id!: string

    @Property({ columnType: 'timestamptz', defaultRaw: 'CURRENT_TIMESTAMP' })
    createdAt = new Date()

    @Property({ columnType: 'timestamptz', defaultRaw: 'CURRENT_TIMESTAMP', onUpdate: () => new Date() })
    updatedAt = new Date()

    // Or here
    @Index("data->>'somefield'")
    @Property({ type: JsonType })
    data: PossibleData

    @Property()
    @Index()
    userId: UserId
}
```
Or something like this
```ts
import { BaseEntity, Entity, Index, JsonType, PrimaryKey, Property } from '@mikro-orm/core'

@Index({ properties: [{data: { somefield: true }}, 'userId'] })
@Entity()
export class Test extends BaseEntity<Test, 'id'> {
    @PrimaryKey({ columnType: 'uuid', defaultRaw: 'uuid_generate_v4()' })
    id!: string

    @Property({ columnType: 'timestamptz', defaultRaw: 'CURRENT_TIMESTAMP' })
    createdAt = new Date()

    @Property({ columnType: 'timestamptz', defaultRaw: 'CURRENT_TIMESTAMP', onUpdate: () => new Date() })
    updatedAt = new Date()

     // Or here
    @Index({ properties: {data: { somefield: true }} })
    @Property({ type: JsonType })
    data: PossibleData

    @Property()
    @Index()
    userId: UserId
}
```

Which when you try to create a migration return 
`MetadataError: Entity Test has wrong index definition: 'data->>'somefield'' does not exist. You need to use property name, not column name.` 

**Describe alternatives you've considered**
Manual migration but mikro-orm try to delete my index with every new migration created, but maybe I missed something

**Comment by B4nan** [comment]
> Manual migration but mikro-orm try to delete my index with every new migration created, but maybe I missed something

A workaround is to define any index with some name, index diffing works just based on index names, so it is enough to have some index with the name and it will be kept intact.

**Comment by B4nan** [comment]
In v5 you can custom index expressions:

```ts
@Index({ name: 'test_on_json_index', expression: `CREATE INDEX "test_on_json_index" ON test((data->>'somefield'), "userId");` })
```

This works both on entity level and property level and you have absolute control over the index definition.

The diffing works again based on index name (that can be either explicitly defined or inferred from the property name). The previous workaround probably won't work anymore as indexes are now properly diffed not just based on name, but also based on fields and their order.

I still see a room for the proposed solution, probably with dot notation instead of objects or platform specific syntax, like in #2129:

```ts
@Index({ properties: ['data.somefield', 'userId'] })
@Entity()
export class Test extends BaseEntity<Test, 'id'> {

    @Index({ properties: 'data.somefield' }) // can be a string too
    @Property({ type: JsonType })
    data: PossibleData

}
```

**Comment by AzSiAz** [comment]
Didn't know about the upcoming v5 change, thanks for the heads up 

The idea with object syntax was to provide autocompletion and more importantly TS type error if something change in the future

**Comment by B4nan** [comment]
> The idea with object syntax was to provide autocompletion and more importantly TS type error if something change in the future

v5 supports type safe dot notation paths, but in general you can't have anything type safe inside decorators unless you explicitly provide the generic argument (it can't be inferred from the class where the decorator is used)

**Comment by AzSiAz** [comment]
Yeah, I was speaking more with using something like 
```ts
@Index({ properties: { data: { field: true | 'asc' | 'desc' } } } as IndexType<JSONType>)
```

But if dot notation in paths is typesafe I'm all in 💯

My only downside from using mikro-orm despite all the amazing work you and contributor put in is type safe usage compared to prisma (still lacking micro-orm flexibility)

**Comment by SteveMelons** [comment]
What would be the current workaround in v5 for nested Index using the mongo driver?

**Comment by SteveMelons** [comment]
> What would be the current workaround in v5 for nested Index using the mongo driver?

Found the solution which does not allow for extra options to be passed as far as I understand though

```
@Index({ options: { "address.location": "2dsphere" } })
```

**Comment by SevenWaysDP** [comment]
> > What would be the current workaround in v5 for nested Index using the mongo driver?
> 
> Found the solution which does not allow for extra options to be passed as far as I understand though
> 
> ```
> @Index({ options: { "address.location": "2dsphere" } })
> ```

It works, on entity level! Thanks!

**Comment by B4nan** [comment]
Closing as implemented in v6 via #4735
